staging/src/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator.go

```diff
@@ -60,24 +61,40 @@ func fmtRawDoc(rawDoc string) string {
 		}
 	}
 
-	// Ignore all lines after ---
-	rawDoc = strings.Split(rawDoc, "---")[0]
+	var isPrevLinePartOfListOrTable bool
+	isPrevLinePartOfListOrTable = false
 
+out:
 	for _, line := range strings.Split(rawDoc, "\n") {
 		line = strings.TrimRight(line, " ")
 		leading := strings.TrimLeft(line, " ")
+		horizontalLineRegex := regexp.MustCompile("^---$")
 		switch {
+		case horizontalLineRegex.Match([]byte(line)): // Ignore all lines after "^---$"
+			delPrevChar()
+			break out
 		case len(line) == 0: // Keep paragraphs
 			delPrevChar()
 			buffer.WriteString("\n\n")
 		case strings.HasPrefix(leading, "TODO"): // Ignore one line TODOs
 		case strings.HasPrefix(leading, "+"): // Ignore instructions to the generators
 		default:
-			if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
+			orderedListRegex := regexp.MustCompile("^[0-9]+\\. ")
+			if strings.HasPrefix(line, "-") || strings.HasPrefix(line, "*") || orderedListRegex.Match([]byte(line)) || strings.HasPrefix(line, "|") { // Unordered/Ordered list items, or table
+				delPrevChar()
+				line = "\n" + line + "\n" // Replace it with newline. This is useful when we have a line with: "Example:\n\tJSON-someting..."
+				isPrevLinePartOfListOrTable = true
+			} else if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
 				delPrevChar()
 				line = "\n" + line + "\n" // Replace it with newline. This is useful when we have a line with: "Example:\n\tJSON-someting..."
+				isPrevLinePartOfListOrTable = true
```

Comment:
I am not sure how this means the line is the start of the table?

Reply:
Thanks for explaining.

I reread this section, and it is a bit hairy. I hope there is a more accessible way to express this code. 

I also think your heuristics which assume any line starting with "\*", "\t", or " " is part of a list or table are questionable. They seem liable to pick up many false positives. For instance, isn't it possible that not all lines starting with " " or "*" are part of a list?

There are unit tests for swagger_doc_generator [here](https://github.com/kubernetes/kubernetes/blob/9054e70ca3a23b0c9fc5e20bafc41b0da081821e/staging/src/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator_test.go#L27-L35). It seems a good place to prove with tests that some of these transformations don't accidentally make undesirable transformations.

Reply:
> I understand a lot better the heuristics you are trying to implement, but I am not convinced the current form isn't risky. I'd like to see quite a lot of tests added to the [unit test framework](https://github.com/kubernetes/kubernetes/blob/9054e70ca3a23b0c9fc5e20bafc41b0da081821e/staging/src/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator_test.go#L27-L35) that demonstrates clearly how this transforms doc comments with a lot of edge cases.

> There are unit tests for swagger_doc_generator [here](https://github.com/kubernetes/kubernetes/blob/9054e70ca3a23b0c9fc5e20bafc41b0da081821e/staging/src/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator_test.go#L27-L35). It seems a good place to prove with tests that some of these transformations don't accidentally make undesirable transformations.

Okay, I will add some unit tests to the `test.go` file.

> I also think your heuristics which assume any line starting with "\*", "\t", or " " is part of a list or table are questionable. They seem liable to pick up many false positives. For instance, isn't it possible that not all lines starting with " " or "\*" are part of a list?

You are right. There exists possibility that not all lines starting with " " or "*" are part of a list.
But,
- My looking over through some `types.go` files could not find cases that invokes the `if` condition.
- Even if some false positives occur, IMO it's okay, because what I wanted to implement here is
  - not 'detecting strictly lists and tables only, and nothing else',
  - but adding missing line breaks (and thus preserving the line breaks/wraps) around some markdown document components including lists and tables,
    and this addition of line breaks does not alter the finally rendered HTML view (e.g. [pod-v1](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/)) of plain text paragraph. An example follows.

Comment in `types.go`:
```go
	// Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse ante dui, 
	// consectetur eget ipsum sed, placerat sodales tellus. Quisque a dictum arcu. Mauris 
	// placerat risus tellus, et luctus massa fringilla a. 
	//   This paragraph is indented with 2 spaces. This paragraph is indented with 
	//   2 spaces. This paragraph is indented with 2 spaces. This paragraph is indented with 2 spaces. 
	// Back again with no indent.
```

Generated `types_swagger_doc_generated.go`:
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse ante dui, consectetur eget ipsum sed, placerat sodales tellus. Quisque a dictum arcu. Mauris placerat risus tellus, et luctus massa fringilla a.\n  This paragraph is indented with 2 spaces. This paragraph is indented with\n  2 spaces. This paragraph is indented with 2 spaces. This paragraph is indented with 2 spaces.\n\nBack again with no indent.
```

Manually beautified, replacing `\n` with actual line break:
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse ante dui, consectetur eget ipsum sed, placerat sodales tellus. Quisque a dictum arcu. Mauris placerat risus tellus, et luctus massa fringilla a.
  This paragraph is indented with 2 spaces. This paragraph is indented with
  2 spaces. This paragraph is indented with 2 spaces. This paragraph is indented with 2 spaces.

Back again with no indent.
```

When this generated result gets rendered with markdown renderer, it looks like:

![image](https://user-images.githubusercontent.com/46767780/207786477-3e877147-3255-42d2-a6dc-2b25ae2266ed.png)
(Screenshot from https://markdownlivepreview.com/)

And we can see that this addition of line breaks does not alter the finally rendered HTML view of plain text paragraph.

Reply:
And if appropriate, we might enhance the condition that catches unordered list
from `HasPrefix(line, "-") || HasPrefix(line, "*")`
to `HasPrefix(line, "- ") || HasPrefix(line, "* ")`.
(Please note that the single spaces are added right after the hyphen and asterisk.)